crates/red_knot_python_semantic/src/semantic_index.rs

```diff
@@ -139,6 +141,11 @@ pub(crate) struct SemanticIndex<'db> {
 
     /// Flags about the global scope (code usage impacting inference)
     has_future_annotations: bool,
+
+    /// Maps from class body scopes to attribute assignments that were found
+    /// in methods of that class.
+    attribute_assignments:
+        FxHashMap<FileScopeId, FxHashMap<&'db str, Vec<AttributeAssignment<'db>>>>,
```

Comment:
I'm unsure if using a `&'db str` here is actually safe... I thought that references is generally disallowed (by having to implement the `Update` trait) but it Salsa seems to allow it here. 

The issue I see here is that the AST might get collected between two revisions (because of LRU caching), in which case the map keys would point to invalid memory addresses. 

I posted in [salsa's zullip](https://salsa.zulipchat.com/#narrow/channel/145099-Using-Salsa/topic/Query.20result.20with.20.26'db.20lifetime.20field) but maybe @carljm remembers

Reply:
Thanks, Micha. I think both Alex and I had hesitations when replacing `String` with `&'db str` here, but we assumed it would be fine when it compiled.

Not this exact `&'db` reference, but I managed to make red_knot segfault today when I returned a `&'db` reference from a query. As noted in the zulip thread, this seems to be a known salsa bug: https://github.com/salsa-rs/salsa/issues/610.

Will revert 59d112f6cb537907e4dd4e10605a881c1254f401

Reply:
You may want to use `Name` instead of `String` to avoid allocations in most cases. 

We may want to explore interning `Name`s... now that we're storing them in many tables.  Or could we store a `FileSymbolId` instead? Ah no, that doesn't work because `x` in `self.x` is not a symbol, only `self` is

Reply:
> I think both Alex and I had hesitations when replacing `String` with `&'db str` here, but we assumed it would be fine when it compiled.

heh, yes!